tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/json/JsonToTlaViaBuilder.scala

```diff
@@ -0,0 +1,274 @@
+package at.forsyte.apalache.tla.typecomp.json
+
+import at.forsyte.apalache.io.json._
+import at.forsyte.apalache.io.lir.TypeTagReader
+import at.forsyte.apalache.tla.imp.src.{SourceLocation, SourceStore}
+import at.forsyte.apalache.tla.types.tla
+import at.forsyte.apalache.tla.typecomp._
+import at.forsyte.apalache.tla.lir.TypedPredefs._
+import at.forsyte.apalache.tla.lir.src.{SourceLocation, SourcePosition, SourceRegion}
+import at.forsyte.apalache.tla.lir._
+
+import scala.util.{Failure, Success, Try}
+
+/**
+ * A JsonDecoder, which uses the ScopedBuilder to construct TLA IR
+ *
+ * TODO: Once package refactoring is done, this should replace io.json.JsonToTla. Currently impossible due to dependency
+ * issues.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+class JsonToTlaViaBuilder[T <: JsonRepresentation](
+    scalaFactory: ScalaFactory[T],
+    sourceStoreOpt: Option[SourceStore] = None,
+  )(implicit typeTagReader: TypeTagReader)
+    extends JsonDecoder[T] {
+  private def missingField(fieldName: String): JsonDeserializationError =
+    new JsonDeserializationError(s"JSON has no field named [$fieldName].")
+
+  private def getOrThrow(json: T, fieldName: String): T =
+    json.getFieldOpt(fieldName).getOrElse { throw missingField(fieldName) }
+
+  private def rethrowAsJsonDeserializationException[T](method: => T): T =
+    try {
+      method
+    } catch {
+      // rethrow as deserialization error
+      case e: TBuilderTypeException =>
+        throw new JsonDeserializationError(s"JSON contains ill-typed expressions: ${e.getMessage}")
+      case e: TBuilderScopeException =>
+        throw new JsonDeserializationError(s"JSON contains ill-scoped expressions: ${e.getMessage}")
```

Comment:
Tracing is easier if we also wrap the causing exception:
https://docs.oracle.com/javase/7/docs/api/java/lang/Exception.html#Exception(java.lang.String,%20java.lang.Throwable)